kart/spatial_tree.py

```diff
@@ -488,64 +396,127 @@ def _find_geometry_column(fields):
     return result
 
 
-def get_s2_tokens(s2_indexer, geom, transforms):
-    envelope = get_envelope_for_indexing(geom, transforms)
-    return envelope_to_s2_tokens(s2_indexer, envelope)
+def get_encoded_envelope(geom, transforms):
+    """
+    Returns an envelope of the geometry in EPSG:4326 (using the given transform(s) to convert it),
+    and encoded as a short string of bytes, ready for writing to the sqlite database.
+    """
+    return encode_envelope(get_envelope_for_indexing(geom, transforms))
+
+
+# Increasing this parameter means the index is more accurate, but takes up more space.
+BITS_PER_VALUE = 20  # Keep this parameter even otherwise you waste half-bytes.
+BITS_PER_ENVELOPE = 4 * BITS_PER_VALUE
+BYTES_PER_ENVELOPE = math.ceil(BITS_PER_ENVELOPE / 8)
+MAX_ENCODED_VALUE = 2 ** BITS_PER_VALUE - 1
+MAX_ENCODED_ENVELOPE = 2 ** BITS_PER_ENVELOPE - 1
+BYTE_ORDER = "big"
+
+
+def encode_envelope(envelope):
+    """
+    Encodes a (min-x, max-x, min-y, max-y) envelope where max-x can be on the far side of the antimeridian.
+    Encodes each value using BITS_PER_VALUE bits, as an unsigned int that exactly spans the entire valid range.
+    """
+    # Note: the x-values have a valid range of 720 degrees at the equator. We could get an extra bit of
+    # precision if we gave them a valid range of only 360 degrees and stored max-x as an offset from min-x, but
+    # it makes the code more complicated.
+    result = _encode_value(envelope[0], -180, 540, math.floor)
+    result <<= BITS_PER_VALUE
+    result |= _encode_value(envelope[1], -180, 540, math.ceil)
+    result <<= BITS_PER_VALUE
+    result |= _encode_value(envelope[2], -90, 90, math.floor)
+    result <<= BITS_PER_VALUE
+    result |= _encode_value(envelope[3], -90, 90, math.ceil)
+    assert 0 <= result <= MAX_ENCODED_ENVELOPE
+    return result.to_bytes(BYTES_PER_ENVELOPE, BYTE_ORDER)
+
+
+def _encode_value(value, min_value, max_value, round_fn):
+    normalised = (value - min_value) / (max_value - min_value)
```

Comment:
These encode a -180° → +540° longitude range, but AFAICS nothing _puts_ coordinates into that range?

Reply:
Yeah - I started at the encoding end (ie, the end result) and had a plan to use this scheme: -180 <= w <= 180, and e - w is the width, where 0 <= width <= 360. But I hadn't got as far as handling antimeridians properly, so I just had a TODO that this is what the numbers should be - the actual envelopes are just whatever OGR or GPKG gives us. I'm adding more anti-meridian handling now in terms of fixing how envelopes are unioned together etc but I still haven't quite gotten all the way to the source - still not quite sure what antimeridian envelopes are like when we pull them out of OGR or GPKG.